### PR TITLE
Disable scheduler with runtime config

### LINKED ISSUE #8913: Scheduler - add config property to disable the scheduler completely
This could be useful for running tests.

### PR DESCRIPTION
**Pull request by adrianfiedler**
Fixes #8913

### CODE REVIEW

File: extensions/quartz/runtime/src/main/java/io/quarkus/quartz/runtime/QuartzRuntimeConfig.java
Comment (mkouba) [inline extensions/quartz/runtime/src/main/java/io/quarkus/quartz/runtime/QuartzRuntimeConfig.java:34]: There is no need to introduce this config property. Quartz extension is built on top of the Scheduler extension - it's ok to reuse `quarkus.scheduler.enabled`.

File: extensions/quartz/deployment/src/test/java/io/quarkus/quartz/test/DisabledSchedulerTest.java
Comment (mkouba) [inline extensions/quartz/deployment/src/test/java/io/quarkus/quartz/test/DisabledSchedulerTest.java:29]: This test will always take at least 5s :-(

File: extensions/quartz/runtime/src/main/java/io/quarkus/quartz/runtime/QuartzScheduler.java
Comment (mkouba) [inline extensions/quartz/runtime/src/main/java/io/quarkus/quartz/runtime/QuartzScheduler.java:72]: As I commented on the issue - the preferred solution is to inject the config group directly.

**Review by mkouba — CHANGES_REQUESTED** [review]

**Comment by mkouba** [comment]
@adrianfiedler A tip for the next time - it's always good to start with a Draft PR unless you're sure the API and implementation is final, functional and ready for review... ;-)

**Comment by adrianfiedler** [comment]
> @adrianfiedler A tip for the next time - it's always good to start with a Draft PR unless you're sure the API and implementation is final, functional and ready for review... ;-)

Ah sorry, didn't even know this exists. I changed it to draft.

**Comment by mkouba** [comment]
@adrianfiedler No problem at all! ;-)

File: extensions/quartz/runtime/src/main/java/io/quarkus/quartz/runtime/QuartzScheduler.java
Comment (adrianfiedler) [reply extensions/quartz/runtime/src/main/java/io/quarkus/quartz/runtime/QuartzScheduler.java]: Ok, I tried to inject the runtime config as a variable into QuartzScheduler. Here the tests failed because it was always null. But then I injected it into the constructor and now the tests pass. 
But if I use the extension directly in any project now, I get those errors with `Unsatisfied dependency for type io.quarkus.scheduler.runtime.SchedulerRuntimeConfig and qualifiers [@Default]`

File: extensions/quartz/deployment/src/test/java/io/quarkus/quartz/test/DisabledSchedulerTest.java
Comment (adrianfiedler) [reply extensions/quartz/deployment/src/test/java/io/quarkus/quartz/test/DisabledSchedulerTest.java]: I reduced the wait time for the latch to 1.1 seconds. Not sure how to get under that time with `@Scheduled(every = "1s")`.

File: extensions/quartz/runtime/src/main/java/io/quarkus/quartz/runtime/QuartzRuntimeConfig.java
Comment (adrianfiedler) [reply extensions/quartz/runtime/src/main/java/io/quarkus/quartz/runtime/QuartzRuntimeConfig.java]: Ok, I removed the configItem from `QuartzRuntimeConfig`. Now there is only one in `SchedulerRuntimeConfig` called `quarkus.scheduler.enabled`.

File: extensions/quartz/runtime/src/main/java/io/quarkus/quartz/runtime/QuartzScheduler.java
Comment (mkouba) [reply extensions/quartz/runtime/src/main/java/io/quarkus/quartz/runtime/QuartzScheduler.java]: > But then I injected it into the constructor and now the tests pass.

That's weird. Constructor and field injection should behave identically. I'll try to change the code locally...

> But if I use the extension directly in any project now...

Does it happen in both the production and the dev mode?

File: extensions/quartz/runtime/src/main/java/io/quarkus/quartz/runtime/QuartzScheduler.java
Comment (mkouba) [inline extensions/quartz/runtime/src/main/java/io/quarkus/quartz/runtime/QuartzScheduler.java:75]: The log message should probably reflect the conditions...

File: extensions/scheduler/runtime/src/main/java/io/quarkus/scheduler/runtime/SimpleScheduler.java
Comment (mkouba) [inline extensions/scheduler/runtime/src/main/java/io/quarkus/scheduler/runtime/SimpleScheduler.java:59]: I think that we should add an INFO log message here as well. WDYT?

File: extensions/quartz/deployment/src/test/java/io/quarkus/quartz/test/DisabledSchedulerTest.java
Comment (mkouba) [reply extensions/quartz/deployment/src/test/java/io/quarkus/quartz/test/DisabledSchedulerTest.java]: Hm, maybe we should add `boolean isRunning()` to the `io.quarkus.scheduler.Scheduler` so that it would be easier to test the property. 

We should also clarify that `Scheduler#pause()` and `Scheduler#resume()` are no-op if the scheduler is disabled. And probably log a warning if it's disabled and someone attempts to call these methods.

File: extensions/quartz/runtime/src/main/java/io/quarkus/quartz/runtime/QuartzScheduler.java
Comment (mkouba) [reply extensions/quartz/runtime/src/main/java/io/quarkus/quartz/runtime/QuartzScheduler.java]: So I've tried to change the code to use field injection (i.e. `@Inject SchedulerRuntimeConfig config`) and it works fine. However, constructor injection is more appropriate in this particular use case (we don't need to store the config in a field ;-).

Comment (adrianfiedler) [reply extensions/quartz/runtime/src/main/java/io/quarkus/quartz/runtime/QuartzScheduler.java]: Thanks for investigating! I guess it is then some local mistake on my side.
Then I will leave it as injected constructor param.

File: extensions/quartz/deployment/src/test/java/io/quarkus/quartz/test/DisabledSchedulerTest.java
Comment (adrianfiedler) [reply extensions/quartz/deployment/src/test/java/io/quarkus/quartz/test/DisabledSchedulerTest.java]: > Hm, maybe we should add `boolean isRunning()` to the `io.quarkus.scheduler.Scheduler` so that it would be easier to test the property.
Ok, should the isRunning value be set by the runtime config value in Scheduler instanciation (so test would basically only check the right config value)? Or should I also check that really no triggers have been created etc.?

File: extensions/quartz/runtime/src/main/java/io/quarkus/quartz/runtime/QuartzScheduler.java
Comment (mkouba) [reply extensions/quartz/runtime/src/main/java/io/quarkus/quartz/runtime/QuartzScheduler.java]: I've also verified that it works with both the dev and the prod mode using the [scheduler-quickstart](https://github.com/quarkusio/quarkus-quickstarts/tree/master/scheduler-quickstart)

File: extensions/quartz/deployment/src/test/java/io/quarkus/quartz/test/DisabledSchedulerTest.java
Comment (mkouba) [reply extensions/quartz/deployment/src/test/java/io/quarkus/quartz/test/DisabledSchedulerTest.java]: I'd say that it should always return `false` if `enabled=false`. Also don't forget that we have 2 impls of `Scheduler`. The default one and the quartz one. We need to think it over ;-).

Comment (adrianfiedler) [reply extensions/quartz/deployment/src/test/java/io/quarkus/quartz/test/DisabledSchedulerTest.java]: Not sure about QuartzScheduler, should we use something like scheduler.getCurrentlyExecutingJobs() > 0 to check if it (any job) is running? Or is a simple boolean check like in SimpleScheduler enough?

Comment (mkouba) [reply extensions/quartz/deployment/src/test/java/io/quarkus/quartz/test/DisabledSchedulerTest.java]: That's a good question. Maybe we should use `org.quartz.Scheduler.isInStandbyMode()` but in that case we should modify the behavior of `QuartzScheduler.pause()` and `QuartzScheduler.resume()`. Currently, we call `pauseAll()/resumeAll()` but we should probably call `standby()/start()` instead. @machi1990 WDYT?

Comment (adrianfiedler) [reply extensions/quartz/deployment/src/test/java/io/quarkus/quartz/test/DisabledSchedulerTest.java]: And maybe another question :) How is it possible to inject the SimpleSchedule into my deployment test to check the isRunning flag? For QuartzScheduler it works because of @Produces.

File: extensions/scheduler/runtime/src/main/java/io/quarkus/scheduler/runtime/SimpleScheduler.java
Comment (machi1990) [inline extensions/scheduler/runtime/src/main/java/io/quarkus/scheduler/runtime/SimpleScheduler.java:168]: nitpick:

```suggestion
    public boolean isRunning() {
        return enabled && running;
    }
```

File: extensions/quartz/deployment/src/test/java/io/quarkus/quartz/test/DisabledSchedulerTest.java
Comment (machi1990) [reply extensions/quartz/deployment/src/test/java/io/quarkus/quartz/test/DisabledSchedulerTest.java]: Let's do it.

File: extensions/quartz/deployment/src/test/java/io/quarkus/quartz/test/DisabledSchedulerTest.java
Comment (machi1990) [inline extensions/quartz/deployment/src/test/java/io/quarkus/quartz/test/DisabledSchedulerTest.java:20]: According to this https://github.com/quarkusio/quarkus/issues/8913#issuecomment-621144357, this should be `null` hence causing a NPE at https://github.com/quarkusio/quarkus/pull/9027/files#diff-b22e0da44e71f897e6899c8e8e9626fbR31 or am I missing something? 
/cc @mkouba

Comment (mkouba) [reply extensions/quartz/deployment/src/test/java/io/quarkus/quartz/test/DisabledSchedulerTest.java:20]: +1

File: extensions/quartz/runtime/src/main/java/io/quarkus/quartz/runtime/QuartzScheduler.java
Comment (mkouba) [inline extensions/quartz/runtime/src/main/java/io/quarkus/quartz/runtime/QuartzScheduler.java:64]: This field should be `final`.

File: extensions/quartz/runtime/src/main/java/io/quarkus/quartz/runtime/QuartzScheduler.java
Comment (mkouba) [inline extensions/quartz/runtime/src/main/java/io/quarkus/quartz/runtime/QuartzScheduler.java:231]: Message? In fact, I'd just throw an `IllegalStateException` instead of a warning. Although it seems that `isInStandbyMode()` never throws an expception...

File: extensions/scheduler/runtime/src/main/java/io/quarkus/scheduler/Scheduler.java
Comment (mkouba) [inline extensions/scheduler/runtime/src/main/java/io/quarkus/scheduler/Scheduler.java:22]: I think that we should be more specific here. Something like "If a scheduler is running the triggers are fired and jobs are executed."

File: extensions/scheduler/runtime/src/main/java/io/quarkus/scheduler/runtime/SimpleScheduler.java
Comment (mkouba) [inline extensions/scheduler/runtime/src/main/java/io/quarkus/scheduler/runtime/SimpleScheduler.java:53]: Make it `final` ;-)

File: extensions/quartz/deployment/src/test/java/io/quarkus/quartz/test/DisabledSchedulerTest.java
Comment (mkouba) [reply extensions/quartz/deployment/src/test/java/io/quarkus/quartz/test/DisabledSchedulerTest.java:20]: So I believe that we should throw IllegalStateException from `io.quarkus.quartz.runtime.QuartzScheduler.produceQuartzScheduler()` if `QuartzScheduler.scheduler` is null. However, in this case we inject `io.quarkus.quartz.runtime.QuartzScheduler` which is OK. @adrianfiedler you should rather inject `io.quarkus.scheduler.Scheduler` which is part of the API (and is implemented by `QuartzScheduler`).

File: extensions/quartz/runtime/src/main/java/io/quarkus/quartz/runtime/QuartzScheduler.java
Comment (mkouba) [inline extensions/quartz/runtime/src/main/java/io/quarkus/quartz/runtime/QuartzScheduler.java:225]: `|| scheduler == null` is IMHO redundant and adds no value.

**Review by mkouba — CHANGES_REQUESTED** [review]

File: extensions/quartz/deployment/src/test/java/io/quarkus/quartz/test/DisabledSchedulerTest.java
Comment (adrianfiedler) [reply extensions/quartz/deployment/src/test/java/io/quarkus/quartz/test/DisabledSchedulerTest.java:20]: Totally agree, how about "Cannot inject not started scheduler" as message for the ISE?"

Comment (mkouba) [reply extensions/quartz/deployment/src/test/java/io/quarkus/quartz/test/DisabledSchedulerTest.java:20]: I would mention Quartz here. Something like "Cannot produce org.quartz.Scheduler - Quartz scheduler is disabled".

Comment (mkouba) [reply extensions/quartz/deployment/src/test/java/io/quarkus/quartz/test/DisabledSchedulerTest.java:20]: Hm, looking at the CDI spec, we should already throw `IllegalProductException` because the producer is not `@Dependent` and returns `null`. Let me check this...

Comment (mkouba) [reply extensions/quartz/deployment/src/test/java/io/quarkus/quartz/test/DisabledSchedulerTest.java:20]: Ah, we only throw `IllegalProductException` for normal scoped producers. Which is probably OK.

Comment (adrianfiedler) [reply extensions/quartz/deployment/src/test/java/io/quarkus/quartz/test/DisabledSchedulerTest.java:20]: There is also one additonal case where org.quartz.Scheduler is null: When no scheduled methods are found and `forceStart `config is `false`.

Comment (mkouba) [reply extensions/quartz/deployment/src/test/java/io/quarkus/quartz/test/DisabledSchedulerTest.java:20]: Good point!

File: extensions/quartz/runtime/src/main/java/io/quarkus/quartz/runtime/QuartzScheduler.java
Comment (adrianfiedler) [reply extensions/quartz/runtime/src/main/java/io/quarkus/quartz/runtime/QuartzScheduler.java:225]: It can be `enabled = true` but scheduler can still be `null` (When no scheduled methods are found and `forceStart` config is false.)

Comment (mkouba) [reply extensions/quartz/runtime/src/main/java/io/quarkus/quartz/runtime/QuartzScheduler.java:225]: You're right. I'm sorry for the confusion...

Comment (adrianfiedler) [reply extensions/quartz/runtime/src/main/java/io/quarkus/quartz/runtime/QuartzScheduler.java:225]: No problem at all, I just wanted to double check in case I oversee something.

File: extensions/quartz/deployment/src/test/java/io/quarkus/quartz/test/DisabledSchedulerTest.java
Comment (adrianfiedler) [reply extensions/quartz/deployment/src/test/java/io/quarkus/quartz/test/DisabledSchedulerTest.java:20]: So because the producer is @Singleton and it throws automatically IllegalProductException - no need to explicitly throw an ISE?

Comment (adrianfiedler) [reply extensions/quartz/deployment/src/test/java/io/quarkus/quartz/test/DisabledSchedulerTest.java:20]: Okay I double checked and injected `org.quartz.Scheduler` directly and did not get an `IllegalProductException`. I will throw then explicitly the ISE.

Comment (mkouba) [reply extensions/quartz/deployment/src/test/java/io/quarkus/quartz/test/DisabledSchedulerTest.java]: Yes, `@Singleton` is not a normal scope. +1 for throwing ISE.

**Comment by adrianfiedler** [comment]
Ok, I think I have addressed all points.

**Review by mkouba — APPROVED** [review]

**Comment by mkouba** [comment]
@adrianfiedler Pls rebase and squash the commits.

File: extensions/quartz/deployment/src/test/java/io/quarkus/quartz/test/DisabledSchedulerTest.java
Comment (machi1990) [inline extensions/quartz/deployment/src/test/java/io/quarkus/quartz/test/DisabledSchedulerTest.java:20]:
```suggestion
    Scheduler scheduler;
```

Comment (machi1990) [reply extensions/quartz/deployment/src/test/java/io/quarkus/quartz/test/DisabledSchedulerTest.java:20]: Also this tests is really like this one https://github.com/quarkusio/quarkus/pull/9027/files#diff-5028b2d8d0d2b040363a8c0981f36caeR17, I think we do not need this.

File: extensions/scheduler/deployment/src/test/java/io/quarkus/scheduler/test/DisabledSchedulerTest.java
Comment (machi1990) [inline extensions/scheduler/deployment/src/test/java/io/quarkus/scheduler/test/DisabledSchedulerTest.java:34]: Since we are only testing if scheduler is running or not, I think we can live off this class a.k.a we do not need it.

File: extensions/quartz/deployment/src/test/java/io/quarkus/quartz/test/DisabledSchedulerTest.java
Comment (adrianfiedler) [reply extensions/quartz/deployment/src/test/java/io/quarkus/quartz/test/DisabledSchedulerTest.java:20]: Not sure, because this class tests the QuartzScheduler (org.quartz.Scheduler) functionality. While the other test is about SimpleScheduler functionality.

Comment (machi1990) [reply extensions/quartz/deployment/src/test/java/io/quarkus/quartz/test/DisabledSchedulerTest.java:20]: Yes, except I am not seeing any quartz related stuff being tested. I think we can remove this test class WDYT? @mkouba. 

Other than that this PR is good to go once all the commits have been squashed.

Comment (mkouba) [reply extensions/quartz/deployment/src/test/java/io/quarkus/quartz/test/DisabledSchedulerTest.java:20]: > Yes, except I am not seeing any quartz related stuff being tested. I think we can remove this test class...

I'd rather keep the test because the injected `io.quarkus.scheduler.Scheduler` is the `io.quarkus.quartz.runtime.QuartzScheduler`. So it tests our quartz integration.

Comment (machi1990) [reply extensions/quartz/deployment/src/test/java/io/quarkus/quartz/test/DisabledSchedulerTest.java:20]: Sorry, I confused it with quartz specific scheduler. Let’s keep it then.

File: extensions/scheduler/deployment/src/test/java/io/quarkus/scheduler/test/DisabledSchedulerTest.java
Comment (mkouba) [reply extensions/scheduler/deployment/src/test/java/io/quarkus/scheduler/test/DisabledSchedulerTest.java:34]: I think that we do need it because if there is no scheduled method then `SimpleScheduler` is never started.

**Comment by adrianfiedler** [comment]
Ok, is now squashed and rebased

**Comment by machi1990** [comment]
> Ok, is now squashed and rebased

Good stuff. Have you seen this comment? https://github.com/quarkusio/quarkus/pull/9027/files#r421516823 I think the `Jobs` class can be removed from both tests.

**Comment by adrianfiedler** [comment]
> Good stuff. Have you seen this comment? https://github.com/quarkusio/quarkus/pull/9027/files#r421516823 I think the `Jobs` class can be removed from both tests.

Ah, I thought this was resolved with the comment from @mkouba after it: https://github.com/quarkusio/quarkus/pull/9027/files#r421678800. 
Because if we remove the `Jobs` classes, the `SimpleScheduler` and the `QuartzScheduler` won't start at all. So we could not really test, that the config works.

**Comment by machi1990** [comment]
> > Good stuff. Have you seen this comment? https://github.com/quarkusio/quarkus/pull/9027/files#r421516823 I think the `Jobs` class can be removed from both tests.
> 
> Ah, I thought this was resolved with the comment from @mkouba after it: https://github.com/quarkusio/quarkus/pull/9027/files#r421678800.
> Because if we remove the `Jobs` classes, the `SimpleScheduler` and the `QuartzScheduler` won't start at all ("No scheduled business methods found"). So we could not really test, that the config works.

My bad, I thought we checked for enablement first: 
- https://github.com/quarkusio/quarkus/pull/9027/files#diff-f3233ae68279acfe2869eaffaeb10682R78 and 
- https://github.com/quarkusio/quarkus/pull/9027/files#diff-dcaa8eebea34f40db3b4dbafda93bac0R61 

but I guess we still need at least one scheduled method to be able to really test the config.

Let's not hold this PR any longer, really good work. Thanks @adrianfiedler.

**Review by machi1990 — APPROVED** [review]

**Comment by adrianfiedler** [comment]
I'm really glad I could contribute :-) Thanks a lot to you two for your patience @mkouba and @machi1990

**Comment by mkouba** [comment]
@adrianfiedler Thanks for your contribution! Looking forward to the next one ;-).

**Pull request merged**
